crates/sui-network/src/discovery/server.rs

```diff
@@ -27,7 +31,32 @@ impl Discovery for Server {
             .our_info
             .clone()
             .ok_or_else(|| anemo::rpc::Status::internal("own_info has not been initialized yet"))?;
-        let known_peers = state.known_peers.values().cloned().collect();
+
+        // prefer returning peers that we are connected to as they are known-good
+        let mut known_peers = state
+            .connected_peers
+            .keys()
+            .filter_map(|peer_id| state.known_peers.get(peer_id))
+            .map(|info| (info.peer_id, info))
+            .take(MAX_PEERS_TO_SEND)
+            .collect::<HashMap<_, _>>();
+
+        if known_peers.len() <= MAX_PEERS_TO_SEND {
```

Comment:
this guarantees that in the case where we have fewer than MAX_PEERS_TO_SEND we will iterate over everything twice - is it worth first checking the size of known_peers and just returning the full contents of that if it's small enough, before going into the selection logic?

Reply:
sure